src/object_store/cache.rs

```diff
@@ -217,88 +223,201 @@ impl CachingObjectStore {
+                chunk_batches.push((key.clone(), tx));
+                // ... and also create a new sender/receiver for the next pending entry
+                (tx, rx) = channel(Ok(Bytes::new()));
+            }
 
-                // While the write task runs cache the in-memory bytes and serve that to all calls
-                // prior to the write task completing.
-                debug!("Caching value for ({path:?}, {chunk}) in memory");
-                Ok(CacheValue::Memory(data))
-            })
-            .await
-            .map_err(|e| object_store::Error::Generic {
-                store: "cache_store",
-                source: Box::new(e),
-            })?;
-
-        match value {
-            CacheValue::File(path, _) => {
-                debug!("Cache value for ({path:?}, {chunk}) fetched from the file");
-                let data = self.file_manager.read_file(path).await.unwrap();
-                Ok(data)
+            if !chunk_batches.is_empty() && (!is_fresh || chunk == end_chunk) {
+                // Evaluate the accumulated chunk batch if either the current entry is not fresh
+                // (meaning we can not extend the batch anymore and keep it continuous), or
+                // we've reached the last chunk (meaning there are not more chunks to be batched).
+                let first = &chunk_batches.first().unwrap().0;
+                let last = &chunk_batches.last().unwrap().0;
+
+                let batch_range = first.range.start..last.range.end;
+                debug!("{path} {chunk_range:?} starting");
+                let mut batch_data = match self.inner.get_range(path, batch_range).await {
+                    Ok(data) => data,
+                    Err(err) => {
+                        debug!("{path} {chunk_range:?} error: {err}");
+                        // object store error is not `Clone`, so we need to do this
+                        let message = format!("{err}");
+                        // Propagate error to all awaiting tasks.
+                        for (_, tx) in chunk_batches {
+                            let _ = tx.send(Err(super::ObjectStoreError::Generic {
+                                reason: message.clone(),
+                            }));
+                        }
+                        return Err(object_store::Error::Generic {
+                            store: "cache_store",
+                            source: Box::new(super::ObjectStoreError::Generic {
+                                reason: message,
+                            }),
+                        });
+                    }
+                };
+                debug!("{path} {chunk_range:?} completed");
+
+                // `Bytes` are `cheaply cloneable and sliceable chunk of contiguous memory`, so
+                // there should be no need to wrap them in an `Arc`.
+                result.put(batch_data.clone());
+
+                // TODO: execute this in a separate task
```

Comment:
If this -> memory -> disk state transition is in a separate task, we don't need another series of task::spawn to do the disk writes, right?

Reply:
We do, since we'd need to join/await on the byte range fetches before returning and don't want to block on the file writes as well.

Reply:
Really? I thought this meant we'd spawn multiple tasks that all have a (key, tx, shared pointer to batch_data) and:

- insert into cache as Memory
- send data to tx
- write to disk
- update in cache as Disk

Spawning these tasks doesn't block us from returning the result of the byte range fetch, right, since the `result` is ready at this point?

Reply:
Ah, fair point (i though the question was spawning the outgoing get_range call as well).

Yeah, that is true this can be spawned off to a separate task.